### ISSUE TITLE
expression-not-assigned / W0106

### ISSUE THREAD

**Issue by da-dada** [PyCQA/pylint#8129]
### Bug description

`def f1():`
`    pass`
`def f2():`
`    pass`
`test = True`
`f1() if test else f2()`

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
test.py:9: [W0106(expression-not-assigned), ] Expression "f1() if test else f2()" is assigned to nothing
```

### Expected behavior

no warning (according to documentation of pylint)

### Pylint version

```shell
newest
```

### OS / Environment

windows

**Comment by mbyrnepr2** [comment]
Thanks for the report @da-dada. This one looks like expected behaviour. Please see the doc for some examples of why this message is triggered: https://pylint.pycqa.org/en/latest/user_guide/messages/warning/expression-not-assigned.html

**Comment by mbyrnepr2** [comment]
Re-opening. Indeed the checker states: Used when an expression **that is not a function call** is assigned to nothing.

**Comment by nickdrozd** [comment]
I think the expression in question is the whole conditional expression `f1() if test else f2()`, which isn't itself a function call. And the value produced by the conditional expression isn't assigned to anything.

**Comment by mbyrnepr2** [comment]
If it is re-written as the following, the message goes away; so I think it makes sense for consistency to consider the one-line version a false positive:
```python
if test:
    f1()
else:
    f2()
```

**Comment by nickdrozd** [comment]
Another way to handle the warning is to explicitly ignore the value:

```python
_ = f1() if test else f2()
```

This acknowledges that a value is being produced and not used.

Switching to an `if / else` block would be the better choice if the purpose of calls is to execute side effects. It's not wrong to use a conditional expression for side effects, but it is a little irregular.

I guess the question is: did they forget to use the value produced by the conditional expression, or did they just want to do a branch as a one-liner?

**Comment by da-dada** [comment]
to get it into perspective I use it to call functions from a library (wxpython) underneath

`wx.CallAfter(self.Layout) if evt_size else self.Layout()`

and they don't give any value back !

discarding a one-liner just to please the linter does not really sound pythonic: I would say it starts to smell..

**Comment by Pierre-Sassoulas** [comment]
We did not detect this false positive in our primer that include pandas, django, numpy, sentry, flask, psychopg, coveragepy, pytest or pygame ..among others. So I would say that calling function conditionally in one liner is not wrong, but pretty unusual for python codebase.

**Comment by da-dada** [comment]
sorry that I didn't consult the 'codebase' (is that synonymous with 'use case' ?)
I'm usually already overwhelmed by the docu, although I think it has improved a lot lately (python I mean)

**Comment by Pierre-Sassoulas** [comment]
Not a reproach, I'm just saying most code bases (i.e. repositories that contain python code) will use this:
```python
if test:
    f1()
else:
    f2()
```
And not the ternary, thus it's the "pythonic" way to do it. We still need to fix the false positive :)

**Comment by nickdrozd** [comment]
But strictly speaking, this isn't a false positive. The conditional expression produces a value, and that value is silently discarded. That's what the checker checks for, and that's what it is. The code isn't erroneous, but it does violate a common convention.

**Comment by da-dada** [comment]
if that was the guideline of the checker then all `f1()` without 'using' the return value should also be flagged

**Comment by mbyrnepr2** [comment]
Looking into the origins of the conditional expression and it looks like `y if x else z` was introduced as a replacement to this syntax: 
`x and y or z`.

I don't see anything specifically stating that it must be assigned to a name; although the real-world code examples in the link mostly are assigning to a name.

[PEP308](https://peps.python.org/pep-0308/)
[Motivating use case](https://mail.python.org/pipermail/python-dev/2005-September/056546.html) (from PEP308)
[real-world code fragments](https://mail.python.org/pipermail/python-dev/2005-September/056803.html) (from PEP308)

**Comment by nickdrozd** [comment]
There are at least four ways to deal with this warning.

1. Convert the conditional expression to an `if / else` block:

```python
# if / else blocks
if test:
    f1()
else:
    f2()
```

2. Explicitly discard the value of the conditional expression:

```python
_ = f1() if test else f2()
```

3. Use a conditional expression to determine the function, then call the result:

```python
(f1 if test else f2)()
```

4. Just disable the warning:

```python
# pylint: disable = expression-not-assigned
f1() if test else f2()
```

If a one-liner is desired, I suggest going with option 3. But in any case I don't think any changes ought to be made to Pylint to accommodate the original. One coder's false positive is another's false negative.

**Comment by Pierre-Sassoulas** [comment]
``1)`` is the code almost everyone use There's only one example of actual function in one line (``print 'succeed' if pred(x) else 'failed', ':', x``) in the link provided by @mbyrnepr2. and it's the recommended snippet since python 3.0 (``I recommend that people leave existing and/or code alone but start writing if/else expressions in new code only.``). But nonetheless it's still valid python code and if the function always return ``None`` then ``2)`` and ``3)`` is just dancing around to appease the linter. Discouraging it should be another message like ``function-call-without-assignement-in-ternary`` and seeing how it's very seldom seen, probably not worth it (imo).

**Comment by da-dada** [comment]
well, if the warning isn't changed (because just main stream is worth caring for) then at least the description shouldn't pretend correctness (just to be honest I think)

**Comment by da-dada** [comment]
finally, to show the elegance of python to the 'real' world (and I think that was the motivation) in gui applications one often has to set attributes to a field conditionally

`self.SetBackgroundColour(wx.RED if enter else wx.BLUE)`

and forcing an assignment in between here would constitute a sacrilege (at least for those writing python, I think)

**Comment by nickdrozd** [comment]
> `self.SetBackgroundColour(wx.RED if enter else wx.BLUE)`

@da-dada How is this example related? Pylint doesn't issue a warning about it because it doesn't discard a value. The value of the conditional expression is immediately passed to a function. No assignment is required, and Pylint doesn't suggest one.

**Comment by da-dada** [comment]
simply as a motivation for a somewhat broader view of the 'code base' & the intension of the **conditional expression** in python (more less nothing of use for pylint)
